pkgs/development/python-modules/fbprophet/default.nix

```diff
@@ -0,0 +1,54 @@
+{
+  stdenv,
+  buildPythonPackage,
+  fetchPypi,
+  cython,
+  pystan,
+  numpy,
+  pandas,
+  matplotlib,
+  lunar-calendar,
+  cmdstanpy,
+  convertdate,
+  holidays,
+  setuptools-git,
+  python-dateutil,
+  setuptools,
+  tqdm,
+  python
+}:
+
+let
+
+  cmdstanpy' = cmdstanpy.overridePythonAttrs (old: rec {
+    pname = "cmdstanpy";
+    version = "0.4.0";
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "07wmv36adp8jxf6zdas8n407dx51q7la486s7s40anvq6q0x412n";
+    };
+  });
```

Comment:
Pinning within python-modules is highly discouraged. This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

The preference to handling this is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

Reply:
I'll make an issue with fbprophet to unpin `cmdstanpy`. In the meantime parhaps I should just pin `cmdstanpy` to `0.4.0` in its nix file? I thought `0.4.0` was super old, but the library actually just on a very fast release cadence. `0.4.0` was released in July 2019.

Reply:
Just found out the master branch of `fbprophet` is using the latest `cmdstanpy` version, so we'll just have to wait for the next release to make it into pypi.

Reply:
if upstream is unlikely to release a new version soon-ish. Then you can always add the patches to enable it, but this might be difficult if there's a large divergence between master and the current released version.